frontend/components/RewardsHistory/useRewardsHistory.ts

```diff
@@ -0,0 +1,124 @@
+  epoch: z.string(),
+  rewards: z.array(z.string()),
+  serviceIds: z.array(z.string()),
+  blockTimestamp: z.string(),
+  transactionHash: z.string(),
+  epochLength: z.number(),
+});
+
+type RewardHistoryResponse = z.infer<typeof RewardHistoryResponseSchema>;
+
+const betaAddress =
+  SERVICE_STAKING_TOKEN_MECH_USAGE_CONTRACT_ADDRESSES[Chain.GNOSIS].pearl_beta;
+const beta2Address =
+  SERVICE_STAKING_TOKEN_MECH_USAGE_CONTRACT_ADDRESSES[Chain.GNOSIS]
+    .pearl_beta_2;
+
+const SUBGRAPH_URL =
+  'https://api.studio.thegraph.com/query/81855/pearl-staking-rewards-history/version/latest';
+
+const fetchRewardsQuery = gql`
+  {
+    allRewards: checkpoints(orderBy: epoch, orderDirection: desc) {
+      blockTimestamp
+      availableRewards
+      epoch
+      epochLength
+      id
+      rewards
+      serviceIds
+      transactionHash
+      contractAddress
+    }
+  }
+`;
+
+const transformRewards = (
+  rewards: RewardHistoryResponse[],
+  serviceId?: number,
+) => {
+  if (!rewards || rewards.length === 0) return [];
+  if (!serviceId) return [];
+
+  return rewards.map((item: RewardHistoryResponse) => {
+    const serviceIdIndex = item.serviceIds.findIndex(
+      (id) => Number(id) === serviceId,
+    );
+    const reward =
+      serviceIdIndex === -1
+        ? 0
+        : ethers.utils.formatUnits(item.rewards[serviceIdIndex], 18);
+
+    return {
+      epochEndTimeStamp: Number(item.blockTimestamp),
+      epochStartTimeStamp:
+        Number(item.blockTimestamp) - Number(item.epochLength),
+      reward: Number(reward),
+      earned: serviceIdIndex !== -1,
+      transactionHash: item.transactionHash,
+    } as EpochDetails;
+  });
+};
+
+export const useRewardsHistory = () => {
+  const { serviceId } = useServices();
+  const { data, isError, isLoading, isFetching, refetch } = useQuery({
+    queryKey: ['rewardsHistory', serviceId],
+    async queryFn() {
+      return await request(SUBGRAPH_URL, fetchRewardsQuery);
+    },
+    select: (data) => {
+      const allRewards = groupBy(
+        (data as { allRewards: RewardHistoryResponse[] }).allRewards,
+        'contractAddress',
+      );
+      const betaRewards = allRewards[betaAddress.toLowerCase()];
+      const beta2Rewards = allRewards[beta2Address.toLowerCase()];
+
+      const rewards = [
+        {
+          id: beta2Address,
+          name: STAKING_PROGRAM_META[StakingProgramId.Beta2].name,
+          history: transformRewards(beta2Rewards, serviceId),
+        },
+        {
+          id: betaAddress,
+          name: STAKING_PROGRAM_META[StakingProgramId.Beta].name,
+          history: transformRewards(betaRewards, serviceId),
+        },
+      ];
+
+      const parsedRewards = StakingRewardSchema.array().safeParse(rewards);
+      if (!parsedRewards.success) {
+        throw new Error(parsedRewards.error.errors.join(', '));
+      }
+
+      return parsedRewards.data;
+    },
+    refetchOnWindowFocus: false,
+    refetchInterval: 5 * 60 * 1000,
```

Comment:
why so small interval? having that the developer query is rate limited, this might cause issues with the server not responding. The epoch lasts ~a day, I suggest either to go with 24h or simply refetch once the rewards are earned. You can do it by removing this key from the cache `queryKey: ['rewardsHistory', serviceId]` [from cache (](https://tanstack.com/query/v4/docs/framework/react/guides/query-invalidation)